src/machine/aot/mod.rs

```diff
@@ -0,0 +1,682 @@
+                        self.labels_to_test.clear();
+                        self.pc = Value::from_u64(next_pc);
+                    }
+                    Err(Error::InvalidInstruction(i)) if i == 0 => {
+                        // Due to alignment or other reasons, the code might
+                        // certain invalid instructions in the executable
+                        // sections, for a normal VM instance that's executing
+                        // instructions, this is usually fine since the invalid
+                        // regions might never be touched. But for an AOT
+                        // solution, this won't work since we have to
+                        // pre-process the whole text section without knowing
+                        // which part would be touched. The solution here, is
+                        // to skip sections containing invalid instructions,
+                        // keep a note of them, and ignore them in the code
+                        // generation phase. One caveat here,
+                        // is that a malicious program might choose to include
+                        // invalid instructions everywhere, hence creating
+                        // numerous sections hoping to bring the program down.
+                        // To tackle that, we will add an upper bound on the
+                        // number of dummy sections allowed here. That
+                        // allow us to signal correct error and revert back
+                        // to assembly VM for those quirky programs.
+                        if !start_of_basic_block {
+                            return Err(Error::OutOfBound);
+                        }
+                        let mut dummy_end = pc + 2;
+                        while dummy_end < section_end && self.memory.execute_load16(dummy_end)? == 0
+                        {
+                            dummy_end += 2;
+                        }
+                        // We checked no sections are overlapped, so dummy
+                        // sections won't overlap with each other as well.
+                        self.dummy_sections.insert(pc, dummy_end);
+                        if self.dummy_sections.len() > MAXIMUM_DUMMY_SECTIONS {
+                            return Err(Error::LimitReached);
+                        }
+                        self.pc = Value::from_u64(dummy_end);
+                    }
+                    Err(e) => return Err(e),
+                }
+            }
+            // A section must end a basic block, otherwise we would run into
+            // out of bound error;
+            if !start_of_basic_block {
+                return Err(Error::OutOfBound);
+            }
+            debug_assert!(!self.labels.contains(&section_end));
+        }
+        // Remove all labels pointed to dummy sections, since we won't generate
+        // code for dummy sections
+        self.labels = self.dummy_sections.iter().fold(
```

Comment:
See `HashSet::retain`

Reply:
This is actually a deliberate choice. If we use `HashSet::retain`, the outer loop would be looping through labels, then for each label we check against all the sections. That means for `n` labels and `m` sections, we need `n*m` checks.

The code here actually has its outer loop looping through sections, then in each section, we loop through remaining labels, that means in each iteration on different sections, we might be dealing with less labels hence less checks.

Note this is not yet benchmarked, it's just due to mental thoughts. I can still change it if you have other thoughts.

Reply:
```
        for (dummy_start, dummy_end) in &self.dummy_sections {
            self.labels.retain(|label| {
                *label < *dummy_start || *label >= *dummy_end
            })
        }
```